README.md
# Elliptic curve cryptography (ECC) in Dart

Elliptic curve cryptography lib for EOS based blockchain in Dart lang.

[![Build Status](https://travis-ci.com/primes-network/eosdart_ecc.svg?branch=master)](https://travis-ci.com/primes-network/eosdart_ecc)


## Usage

A simple usage example:

```dart
import 'package:eosdart_ecc/eosdart_ecc.dart';

main() {
  // Construct the EOS private key from string
  EOSPrivateKey privateKey = EOSPrivateKey.fromString(
      '5KQwrPbwdL6PhXujxW37FSSQZ1JiwsST4cqQzDeyXtP79zkvFD3');

  // Get the related EOS public key
  EOSPublicKey publicKey = privateKey.toEOSPublicKey();
  // Print the EOS public key
  print(publicKey.toString());

  // Going to sign the data
  String data = 'data';

  // Sign
  EOSSignature signature = privateKey.signString(data);
  // Print the EOS signature
  print(signature.toString());

  // Verify the data using the signature
  signature.verify(data, publicKey);
}
```

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

## References

eosjs-ecc: https://github.com/EOSIO/eosjs-ecc

[tracker]: https://github.com/primes-network/eosdart_ecc/issues


CHANGELOG.md
## 0.4.4

- Null safety

## 0.4.3

- Clean up

## 0.4.2

- Support recover public key from signature

## 0.4.1

- Adopt Flutter Web

## 0.4.0

- fix signature bug

## 0.3.2

- fix signature bug

## 0.3.1

- sign method will sign Uint8List
- add signString method which signs String

## 0.3.0

- Support canonical signature

## 0.2.2

- Healthy change

## 0.2.1

- Add CHANGELOG.md

## 0.2.0

- EOSPublicKey, EOSPrivateKey, EOSSignature
- Sign the data
- Verify the signature

## 0.1.0

- Initial version, created by Stagehand


test/key_test.dart
import 'package:eosdart_ecc/eosdart_ecc.dart';
import 'package:test/test.dart';

void main() {
  group('EOS Key tests', () {
    test('Construct EOS public key from string', () {
      EOSPublicKey publicKey = EOSPublicKey.fromString(
          'EOS8Qi58kbERkTJC7A4gabxYU4SbrAxStJHacoke4sf6AvJyEDZXj');
      print(publicKey);

      expect('EOS8Qi58kbERkTJC7A4gabxYU4SbrAxStJHacoke4sf6AvJyEDZXj',
          publicKey.toString());
    });

    test('Construct EOS public key from string PUB_K1 format', () {
      EOSPublicKey publicKey = EOSPublicKey.fromString(
          'PUB_K1_859gxfnXyUriMgUeThh1fWv3oqcpLFyHa3TfFYC4PK2Ht7beeX');
      print(publicKey);
    });

    test('Construct EOS private key from string', () {
      // common private key
      EOSPrivateKey privateKey = EOSPrivateKey.fromString(
          '5J9b3xMkbvcT6gYv2EpQ8FD4ZBjgypuNKwE1jxkd7Wd1DYzhk88');
      expect('EOS8Qi58kbERkTJC7A4gabxYU4SbrAxStJHacoke4sf6AvJyEDZXj',
          privateKey.toEOSPublicKey().toString());
      expect('5J9b3xMkbvcT6gYv2EpQ8FD4ZBjgypuNKwE1jxkd7Wd1DYzhk88',
          privateKey.toString());
    });

    test('Invalid EOS private key', () {
      try {
        EOSPrivateKey.fromString(
            '5KYZdUEo39z3FPrtuX2QbbwGnNP5zTd7yyr2SC1j299sBCnWjsm');
        fail('Should be invalid private key');
      } on InvalidKey {} catch (e) {
        fail('Should throw InvalidKey exception');
      }
    });

    test('Construct random EOS private key from seed', () {
      EOSPrivateKey privateKey = EOSPrivateKey.fromSeed('abc');
      print(privateKey);
      print(privateKey.toEOSPublicKey());

      EOSPrivateKey privateKey2 =
          EOSPrivateKey.fromString(privateKey.toString());
      expect(privateKey.toEOSPublicKey().toString(),
          privateKey2.toEOSPublicKey().toString());
    });

    test('Construct random EOS private key', () {
      EOSPrivateKey privateKey = EOSPrivateKey.fromRandom();

      print(privateKey);
      print(privateKey.toEOSPublicKey());

      EOSPrivateKey privateKey2 =
          EOSPrivateKey.fromString(privateKey.toString());
      expect(privateKey.toEOSPublicKey().toString(),
          privateKey2.toEOSPublicKey().toString());
    });

    test('Construct EOS private key from string in PVT format', () {
      // PVT private key
      EOSPrivateKey privateKey2 = EOSPrivateKey.fromString(
          'PVT_K1_2jH3nnhxhR3zPUcsKaWWZC9ZmZAnKm3GAnFD1xynGJE1Znuvjd');
      print(privateKey2);
    });

    test('Construct EOS private key from string with compress flag', () {
      // Compressed private key
      EOSPrivateKey privateKey3 = EOSPrivateKey.fromString(
          'L5TCkLizyYqjvKSy6jg1XM3Lc4uTDwwvHS2BYatyXSyoS8T5kC2z');
      print(privateKey3);
    });
  });
}


test/signature_test.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:eosdart_ecc/eosdart_ecc.dart';
import 'package:test/test.dart';
import 'package:crypto/crypto.dart';

void main() {
  group('EOS signature tests', () {
    test('Construct EOS signature from string', () {
      String sigStr =
          'SIG_K1_Kg417TSLuhzSpU2bGa21kD1UNaTfAZSCcKmKpZ6fnx3Nqu22gzG3ND4Twur7bzX8oS1J91JvV4rMJcFycGqFBSaY2SJcEQ';
      EOSSignature signature = EOSSignature.fromString(sigStr);
      print(signature);

      expect(sigStr, signature.toString());
    });

    test('Sign the hash using private key', () {
      EOSPrivateKey privateKey = EOSPrivateKey.fromString(
          '5KQwrPbwdL6PhXujxW37FSSQZ1JiwsST4cqQzDeyXtP79zkvFD3');
      EOSPublicKey publicKey = privateKey.toEOSPublicKey();
      String expectedSig =
          'SIG_K1_Kg417TSLuhzSpU2bGa21kD1UNaTfAZSCcKmKpZ6fnx3Nqu22gzG3ND4Twur7bzX8oS1J91JvV4rMJcFycGqFBSaY2SJcEQ';

      String data = 'data';
      Uint8List hashData = sha256.convert(utf8.encode(data)).bytes as Uint8List;
      EOSSignature signature = privateKey.signHash(hashData);
      EOSSignature signature2 = privateKey.signString(data);

      print(signature.toString());
      expect(expectedSig, signature.toString());
      expect(true, signature.verifyHash(hashData, publicKey));
      expect(true, signature2.verifyHash(hashData, publicKey));

      expect(true, signature.verify(data, publicKey));
      expect(true, signature2.verify(data, publicKey));
    });

    test('Sign the hash using private key', () {
      EOSPrivateKey privateKey = EOSPrivateKey.fromString(
          '5HxT6prWB8VuXkoAaX3eby8bWjquMtCvGuakhC8tGEiPSHfsQLR');
      EOSPublicKey publicKey = privateKey.toEOSPublicKey();
      String expectedSig =
          'SIG_K1_Kdfe9wknSAKBmgwb3L53CG8KosoHhZ69oVEJrrH5YuWx4JVcJdn1ZV3MU25AVho4mPbeSKW79DVTBAAWj7zGbHTByF1JXU';

      List<int> l = [
        244,
        163,
        240,
        75,
        174,
        150,
        233,
        185,
        227,
        66,
        27,
        130,
        230,
        139,
        102,
        112,
        128,
        38,
        78,
        233,
        105,
        59,
        61,
        11,
        25,
        221,
        42,
        109,
        80,
        184,
        174,
        201
      ];
      Uint8List hashData = Uint8List.fromList(l);
      EOSSignature signature = privateKey.signHash(hashData);

      expect(expectedSig, signature.toString());
      print(signature.toString());
      expect(true, signature.verifyHash(hashData, publicKey));
    });

    test('Sign the hash using private key', () {
      EOSPrivateKey privateKey = EOSPrivateKey.fromString(
          '5J9b3xMkbvcT6gYv2EpQ8FD4ZBjgypuNKwE1jxkd7Wd1DYzhk88');
      EOSPublicKey publicKey = privateKey.toEOSPublicKey();
      String expectedSig =
          'SIG_K1_KWfDGxwogny1PUiBAYTfKwPsCSNvM7zWgmXyChdYayZFfyPjddpBUYVdJTq1PjC3PRXADRsqWVU1N2SMQivBDqA7AaRzmB';

      List<int> l = [
        136,
        139,
        63,
        11,
        114,
        68,
        227,
        116,
        92,
        61,
        64,
        121,
        147,
        210,
        233,
        25,
        74,
        164,
        140,
        112,
        45,
        5,
        254,
        165,
        208,
        158,
        53,
        212,
        128,
        190,
        153,
        142
      ];
      Uint8List hashData = Uint8List.fromList(l);
      EOSSignature signature = privateKey.signHash(hashData);

      expect(expectedSig, signature.toString());
      print(signature.toString());
      expect(true, signature.verifyHash(hashData, publicKey));
    });

    test('Recover EOSPublicKey from sign data', () {
      const data = 'this is some data to sign';

      var eosPrivateKey = EOSPrivateKey.fromRandom();
      var eosPublicKey = eosPrivateKey.toEOSPublicKey();

      var signature = eosPrivateKey.signString(data);

      var recoveredEOSPublicKey = signature.recover(data);

      expect(eosPublicKey.toString(), recoveredEOSPublicKey.toString());
      print('Generated EOSPublicKey : ${eosPublicKey.toString()}');
      print('Recovered EOSPublicKey : ${recoveredEOSPublicKey.toString()}');
    });
  });
}


test/eosdart_ecc_test.dart
import 'package:eosdart_ecc/eosdart_ecc.dart';
import 'package:test/test.dart';

void main() {
  group('EOS Key tests', () {
    test('Construct EOS public key from string', () {
      EOSPublicKey publicKey = EOSPublicKey.fromString(
          'EOS8Qi58kbERkTJC7A4gabxYU4SbrAxStJHacoke4sf6AvJyEDZXj');
      print(publicKey);

      expect('EOS8Qi58kbERkTJC7A4gabxYU4SbrAxStJHacoke4sf6AvJyEDZXj',
          publicKey.toString());
    });

    test('Construct EOS public key from string PUB_K1 format', () {
      EOSPublicKey publicKey = EOSPublicKey.fromString(
          'PUB_K1_859gxfnXyUriMgUeThh1fWv3oqcpLFyHa3TfFYC4PK2Ht7beeX');
      print(publicKey);
    });

    test('Construct EOS private key from string', () {
      // common private key
      EOSPrivateKey privateKey = EOSPrivateKey.fromString(
          '5J9b3xMkbvcT6gYv2EpQ8FD4ZBjgypuNKwE1jxkd7Wd1DYzhk88');
      expect('EOS8Qi58kbERkTJC7A4gabxYU4SbrAxStJHacoke4sf6AvJyEDZXj',
          privateKey.toEOSPublicKey().toString());
      expect('5J9b3xMkbvcT6gYv2EpQ8FD4ZBjgypuNKwE1jxkd7Wd1DYzhk88',
          privateKey.toString());
    });

    test('Invalid EOS private key', () {
      try {
        EOSPrivateKey.fromString(
            '5KYZdUEo39z3FPrtuX2QbbwGnNP5zTd7yyr2SC1j299sBCnWjsm');
        fail('Should be invalid private key');
      } on InvalidKey {} catch (e) {
        fail('Should throw InvalidKey exception');
      }
    });

    test('Construct random EOS private key from seed', () {
      EOSPrivateKey privateKey = EOSPrivateKey.fromSeed('abc');
      print(privateKey);
      print(privateKey.toEOSPublicKey());

      EOSPrivateKey privateKey2 =
      EOSPrivateKey.fromString(privateKey.toString());
      expect(privateKey.toEOSPublicKey().toString(),
          privateKey2.toEOSPublicKey().toString());
    });

    test('Construct random EOS private key', () {
      EOSPrivateKey privateKey = EOSPrivateKey.fromRandom();

      print(privateKey);
      print(privateKey.toEOSPublicKey());

      EOSPrivateKey privateKey2 =
      EOSPrivateKey.fromString(privateKey.toString());
      expect(privateKey.toEOSPublicKey().toString(),
          privateKey2.toEOSPublicKey().toString());
    });

    test('Construct EOS private key from string in PVT format', () {
      // PVT private key
      EOSPrivateKey privateKey2 = EOSPrivateKey.fromString(
          'PVT_K1_2jH3nnhxhR3zPUcsKaWWZC9ZmZAnKm3GAnFD1xynGJE1Znuvjd');
      print(privateKey2);
    });

    test('Construct EOS private key from string with compress flag', () {
      // Compressed private key
      EOSPrivateKey privateKey3 = EOSPrivateKey.fromString(
          'L5TCkLizyYqjvKSy6jg1XM3Lc4uTDwwvHS2BYatyXSyoS8T5kC2z');
      print(privateKey3);
    });
  });
}


lib/src/key.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';

import 'package:crypto/crypto.dart';
import 'package:pointycastle/ecc/api.dart' show ECSignature, ECPoint;
import 'package:pointycastle/src/utils.dart';

import './exception.dart';
import './key_base.dart';
import './signature.dart';

/// EOS Public Key
class EOSPublicKey extends EOSKey {
  ECPoint? q;

  /// Construct EOS public key from buffer
  EOSPublicKey.fromPoint(this.q);

  /// Construct EOS public key from string
  factory EOSPublicKey.fromString(String keyStr) {
    RegExp publicRegex = RegExp(r"^PUB_([A-Za-z0-9]+)_([A-Za-z0-9]+)",
        caseSensitive: true, multiLine: false);
    Iterable<Match> match = publicRegex.allMatches(keyStr);

    if (match.isEmpty) {
      RegExp eosRegex = RegExp(r"^EOS", caseSensitive: true, multiLine: false);
      if (!eosRegex.hasMatch(keyStr)) {
        throw InvalidKey("No leading EOS");
      }
      String publicKeyStr = keyStr.substring(3);
      Uint8List buffer = EOSKey.decodeKey(publicKeyStr);
      return EOSPublicKey.fromBuffer(buffer);
    } else if (match.length == 1) {
      Match m = match.first;
      String? keyType = m.group(1);
      Uint8List buffer = EOSKey.decodeKey(m.group(2)!, keyType);
      return EOSPublicKey.fromBuffer(buffer);
    } else {
      throw InvalidKey('Invalid public key format');
    }
  }

  factory EOSPublicKey.fromBuffer(Uint8List buffer) {
    ECPoint? point = EOSKey.secp256k1.curve.decodePoint(buffer);
    return EOSPublicKey.fromPoint(point);
  }

  Uint8List toBuffer() {
    // always compressed
    return q!.getEncoded(true);
  }

  String toString() {
    return 'EOS' + EOSKey.encodeKey(this.toBuffer(), keyType);
  }
}

/// EOS Private Key
class EOSPrivateKey extends EOSKey {
  Uint8List? d;
  String? format;

  late BigInt _r;
  late BigInt _s;

  /// Constructor EOS private key from the key buffer itself
  EOSPrivateKey.fromBuffer(this.d);

  /// Construct the private key from string
  /// It can come from WIF format for PVT format
  EOSPrivateKey.fromString(String keyStr) {
    RegExp privateRegex = RegExp(r"^PVT_([A-Za-z0-9]+)_([A-Za-z0-9]+)",
        caseSensitive: true, multiLine: false);
    Iterable<Match> match = privateRegex.allMatches(keyStr);

    if (match.isEmpty) {
      format = 'WIF';
      keyType = 'K1';
      // WIF
      Uint8List keyWLeadingVersion = EOSKey.decodeKey(keyStr, EOSKey.SHA256X2);
      int version = keyWLeadingVersion.first;
      if (EOSKey.VERSION != version) {
        throw InvalidKey("version mismatch");
      }

      d = keyWLeadingVersion.sublist(1, keyWLeadingVersion.length);
      if (d!.lengthInBytes == 33 && d!.elementAt(32) == 1) {
        // remove compression flag
        d = d!.sublist(0, 32);
      }

      if (d!.lengthInBytes != 32) {
        throw InvalidKey('Expecting 32 bytes, got ${d!.length}');
      }
    } else if (match.length == 1) {
      format = 'PVT';
      Match m = match.first;
      keyType = m.group(1);
      d = EOSKey.decodeKey(m.group(2)!, keyType);
    } else {
      throw InvalidKey('Invalid Private Key format');
    }
  }

  /// Generate EOS private key from seed. Please note: This is not random!
  /// For the given seed, the generated key would always be the same
  factory EOSPrivateKey.fromSeed(String seed) {
    Digest s = sha256.convert(utf8.encode(seed));
    return EOSPrivateKey.fromBuffer(Uint8List.fromList(s.bytes));
  }

  /// Generate the random EOS private key
  factory EOSPrivateKey.fromRandom() {
//    final int randomLimit = 1 << 32;
    final int randomLimit = 4294967296;
    Random randomGenerator;
    try {
      randomGenerator = Random.secure();
    } catch (e) {
      randomGenerator = Random();
    }

    int randomInt1 = randomGenerator.nextInt(randomLimit);
    Uint8List entropy1 = encodeBigInt(BigInt.from(randomInt1));

    int randomInt2 = randomGenerator.nextInt(randomLimit);
    Uint8List entropy2 = encodeBigInt(BigInt.from(randomInt2));

    int randomInt3 = randomGenerator.nextInt(randomLimit);
    Uint8List entropy3 = encodeBigInt(BigInt.from(randomInt3));

    List<int> entropy = entropy1.toList();
    entropy.addAll(entropy2);
    entropy.addAll(entropy3);
    Uint8List randomKey = Uint8List.fromList(entropy);
    Digest d = sha256.convert(randomKey);
    return EOSPrivateKey.fromBuffer(Uint8List.fromList(d.bytes));
  }

  /// Check if the private key is WIF format
  bool isWIF() => this.format == 'WIF';

  /// Get the public key string from this private key
  EOSPublicKey toEOSPublicKey() {
    BigInt privateKeyNum = decodeBigIntWithSign(1, this.d!);
    ECPoint? ecPoint = EOSKey.secp256k1.G * privateKeyNum;

    return EOSPublicKey.fromPoint(ecPoint);
  }

  /// Sign the bytes data using the private key
  EOSSignature sign(Uint8List data) {
    Digest d = sha256.convert(data);
    return signHash(Uint8List.fromList(d.bytes));
  }

  /// Sign the string data using the private key
  EOSSignature signString(String data) {
    return sign(Uint8List.fromList(utf8.encode(data)));
  }

  /// Sign the SHA256 hashed data using the private key
  EOSSignature signHash(Uint8List sha256Data) {
    int nonce = 0;
    BigInt n = EOSKey.secp256k1.n;
    BigInt e = decodeBigIntWithSign(1, sha256Data);

    while (true) {
      _deterministicGenerateK(sha256Data, this.d!, e, nonce++);
      var N_OVER_TWO = n >> 1;
      if (_s.compareTo(N_OVER_TWO) > 0) {
        _s = n - _s;
      }
      ECSignature sig = ECSignature(_r, _s);

      Uint8List der = EOSSignature.ecSigToDER(sig);

      int lenR = der.elementAt(3);
      int lenS = der.elementAt(5 + lenR);
      if (lenR == 32 && lenS == 32) {
        int i = EOSSignature.calcPubKeyRecoveryParam(
            decodeBigIntWithSign(1, sha256Data), sig, this.toEOSPublicKey());
        i += 4; // compressed
        i += 27; // compact  //  24 or 27 :( forcing odd-y 2nd key candidate)
        return EOSSignature(i, sig.r, sig.s);
      }
    }
  }

  String toString() {
    List<int> version = <int>[];
    version.add(EOSKey.VERSION);
    Uint8List keyWLeadingVersion =
        EOSKey.concat(Uint8List.fromList(version), this.d!);

    return EOSKey.encodeKey(keyWLeadingVersion, EOSKey.SHA256X2);
  }

  BigInt _deterministicGenerateK(
      Uint8List hash, Uint8List x, BigInt e, int nonce) {
    List<int> newHash = hash;
    if (nonce > 0) {
      List<int> addition = Uint8List(nonce);
      List<int> data = List.from(hash)..addAll(addition);
      newHash = sha256.convert(data).bytes;
    }

    // Step B
    Uint8List v = Uint8List(32);
    for (int i = 0; i < v.lengthInBytes; i++) {
      v[i] = 1;
    }

    // Step C
    Uint8List k = Uint8List(32);

    // Step D
    List<int> d1 = List.from(v)
      ..add(0)
      ..addAll(x)
      ..addAll(newHash);

    Hmac hMacSha256 = Hmac(sha256, k); // HMAC-SHA256
    k = Uint8List.fromList(hMacSha256.convert(d1).bytes);

    // Step E
    hMacSha256 = Hmac(sha256, k); // HMAC-SHA256
    v = Uint8List.fromList(hMacSha256.convert(v).bytes);

    // Step F
    List<int> d2 = List.from(v)
      ..add(1)
      ..addAll(x)
      ..addAll(newHash);

    k = Uint8List.fromList(hMacSha256.convert(d2).bytes);

    // Step G
    hMacSha256 = Hmac(sha256, k); // HMAC-SHA256
    v = Uint8List.fromList(hMacSha256.convert(v).bytes);
    // Step H1/H2a, again, ignored as tlen === qlen (256 bit)
    // Step H2b again
    v = Uint8List.fromList(hMacSha256.convert(v).bytes);

    BigInt T = decodeBigIntWithSign(1, v);
    // Step H3, repeat until T is within the interval [1, n - 1]
    while (T.sign <= 0 ||
        T.compareTo(EOSKey.secp256k1.n) >= 0 ||
        !_checkSig(e, Uint8List.fromList(newHash), T)) {
      List<int> d3 = List.from(v)..add(0);
      k = Uint8List.fromList(hMacSha256.convert(d3).bytes);
      hMacSha256 = Hmac(sha256, k); // HMAC-SHA256
      v = Uint8List.fromList(hMacSha256.convert(v).bytes);
      // Step H1/H2a, again, ignored as tlen === qlen (256 bit)
      // Step H2b again
      v = Uint8List.fromList(hMacSha256.convert(v).bytes);

      T = decodeBigIntWithSign(1, v);
    }
    return T;
  }

  bool _checkSig(BigInt e, Uint8List hash, BigInt k) {
    BigInt n = EOSKey.secp256k1.n;
    ECPoint Q = (EOSKey.secp256k1.G * k)!;

    if (Q.isInfinity) {
      return false;
    }

    _r = Q.x!.toBigInteger()! % n;
    if (_r.sign == 0) {
      return false;
    }

    _s = k.modInverse(EOSKey.secp256k1.n) *
        (e + decodeBigIntWithSign(1, d!) * _r) %
        n;
    if (_s.sign == 0) {
      return false;
    }

    return true;
  }
}


lib/src/key_base.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:bs58check/bs58check.dart';
import 'package:crypto/crypto.dart';
import 'package:pointycastle/digests/ripemd160.dart';
import "package:pointycastle/ecc/curves/secp256k1.dart";
import 'package:pointycastle/src/utils.dart';

import './exception.dart';

/// abstract EOS Key
abstract class EOSKey {
  static final String SHA256X2 = 'sha256x2';
  static final int VERSION = 0x80;
  static final ECCurve_secp256k1 secp256k1 = ECCurve_secp256k1();

  String? keyType;

  /// Decode key from string format
  static Uint8List decodeKey(String keyStr, [String? keyType]) {
    Uint8List buffer = base58.decode(keyStr);

    Uint8List checksum = buffer.sublist(buffer.length - 4, buffer.length);
    Uint8List key = buffer.sublist(0, buffer.length - 4);

    Uint8List newChecksum;
    if (keyType == SHA256X2) {
      newChecksum = sha256x2(key).sublist(0, 4);
    } else {
      Uint8List check = key;
      if (keyType != null) {
        check = concat(key, Uint8List.fromList(utf8.encode(keyType)));
      }
      newChecksum = RIPEMD160Digest().process(check).sublist(0, 4);
    }
    if (decodeBigIntWithSign(1, checksum) !=
        decodeBigIntWithSign(1, newChecksum)) {
      throw InvalidKey("checksum error");
    }
    return key;
  }

  /// Encode key to string format using base58 encoding
  static String encodeKey(Uint8List key, [String? keyType]) {
    if (keyType == SHA256X2) {
      Uint8List checksum = sha256x2(key).sublist(0, 4);
      return base58.encode(concat(key, checksum));
    }

    Uint8List keyBuffer = key;
    if (keyType != null) {
      keyBuffer = concat(key, Uint8List.fromList(utf8.encode(keyType)));
    }
    Uint8List checksum = RIPEMD160Digest().process(keyBuffer).sublist(0, 4);
    return base58.encode(concat(key, checksum));
  }

  /// Do SHA256 hash twice on the given data
  static Uint8List sha256x2(Uint8List data) {
    Digest d1 = sha256.convert(data);
    Digest d2 = sha256.convert(d1.bytes);
    return Uint8List.fromList(d2.bytes);
  }

  static Uint8List concat(Uint8List p1, Uint8List p2) {
    List<int> keyList = p1.toList();
    keyList.addAll(p2);
    return Uint8List.fromList(keyList);
  }

  static List<int> toSigned(Uint8List bytes) {
    List<int> result = <int>[];
    for (int i = 0; i < bytes.length; i++) {
      int v = bytes[i].toSigned(8);
      //TODO I don't know why, just guess...
      if (i == 0 && v < 0) {
        result.add(0);
      }
      result.add(v);
    }
    return result;
  }
}


lib/src/eosdart_ecc_base.dart
import 'dart:typed_data';
import 'dart:convert';
import 'dart:math';

import 'package:crypto/crypto.dart';
import 'package:pointycastle/src/utils.dart';
import "package:pointycastle/api.dart" show PrivateKeyParameter;
import 'package:pointycastle/ecc/api.dart'
    show ECPrivateKey, ECSignature, ECPoint;
import "package:pointycastle/signers/ecdsa_signer.dart";
import 'package:pointycastle/macs/hmac.dart';
import "package:pointycastle/digests/sha256.dart";

import './exception.dart';
import './key_base.dart';
import './signature.dart';

/// EOS Public Key
class EOSPublicKey extends EOSKey {
  ECPoint q;

  /// Construct EOS public key from buffer
  EOSPublicKey.fromPoint(this.q);

  /// Construct EOS public key from string
  factory EOSPublicKey.fromString(String keyStr) {
    RegExp publicRegex = RegExp(r"^PUB_([A-Za-z0-9]+)_([A-Za-z0-9]+)",
        caseSensitive: true, multiLine: false);
    Iterable<Match> match = publicRegex.allMatches(keyStr);

    if (match.length == 0) {
      RegExp eosRegex = RegExp(r"^EOS", caseSensitive: true, multiLine: false);
      if (!eosRegex.hasMatch(keyStr)) {
        throw InvalidKey("No leading EOS");
      }
      String publicKeyStr = keyStr.substring(3);
      Uint8List buffer = EOSKey.decodeKey(publicKeyStr);
      return EOSPublicKey.fromBuffer(buffer);
    } else if (match.length == 1) {
      Match m = match.first;
      String keyType = m.group(1);
      Uint8List buffer = EOSKey.decodeKey(m.group(2), keyType);
      return EOSPublicKey.fromBuffer(buffer);
    } else {
      throw InvalidKey('Invalid public key format');
    }
  }

  factory EOSPublicKey.fromBuffer(Uint8List buffer) {
    ECPoint point = EOSKey.secp256k1.curve.decodePoint(buffer);
    return EOSPublicKey.fromPoint(point);
  }

  Uint8List toBuffer() {
    // always compressed
    return q.getEncoded(true);
  }

  String toString() {
    return 'EOS' + EOSKey.encodeKey(this.toBuffer(), keyType);
  }
}

/// EOS Private Key
class EOSPrivateKey extends EOSKey {
  Uint8List d;
  String format;

  /// Constructor EOS private key from the key buffer itself
  EOSPrivateKey.fromBuffer(this.d);

  /// Construct the private key from string
  /// It can come from WIF format for PVT format
  EOSPrivateKey.fromString(String keyStr) {
    RegExp privateRegex = RegExp(r"^PVT_([A-Za-z0-9]+)_([A-Za-z0-9]+)",
        caseSensitive: true, multiLine: false);
    Iterable<Match> match = privateRegex.allMatches(keyStr);

    if (match.length == 0) {
      format = 'WIF';
      keyType = 'K1';
      // WIF
      Uint8List keyWLeadingVersion = EOSKey.decodeKey(keyStr, EOSKey.SHA256X2);
      int version = keyWLeadingVersion.first;
      if (EOSKey.VERSION != version) {
        throw InvalidKey("version mismatch");
      }

      d = keyWLeadingVersion.sublist(1, keyWLeadingVersion.length);
      if (d.lengthInBytes == 33 && d.elementAt(32) == 1) {
        // remove compression flag
        d = d.sublist(0, 32);
      }

      if (d.lengthInBytes != 32) {
        throw InvalidKey('Expecting 32 bytes, got ${d.length}');
      }
    } else if (match.length == 1) {
      format = 'PVT';
      Match m = match.first;
      keyType = m.group(1);
      d = EOSKey.decodeKey(m.group(2), keyType);
    } else {
      throw InvalidKey('Invalid Private Key format');
    }
  }

  /// Generate EOS private key from seed. Please note: This is not random!
  /// For the given seed, the generated key would always be the same
  factory EOSPrivateKey.fromSeed(String seed) {
    Digest s = sha256.convert(utf8.encode(seed));
    return EOSPrivateKey.fromBuffer(s.bytes);
  }

  /// Generate the random EOS private key
  factory EOSPrivateKey.fromRandom() {
    final int randomLimit = 1 << 32;
    Random randomGenerator;
    try {
      randomGenerator = Random.secure();
    } catch (e) {
      randomGenerator = new Random();
    }

    int randomInt1 = randomGenerator.nextInt(randomLimit);
    Uint8List entropy1 = encodeBigInt(BigInt.from(randomInt1));

    int randomInt2 = randomGenerator.nextInt(randomLimit);
    Uint8List entropy2 = encodeBigInt(BigInt.from(randomInt2));

    int randomInt3 = randomGenerator.nextInt(randomLimit);
    Uint8List entropy3 = encodeBigInt(BigInt.from(randomInt3));

    List<int> entropy = entropy1.toList();
    entropy.addAll(entropy2);
    entropy.addAll(entropy3);
    Uint8List randomKey = Uint8List.fromList(entropy);
    Digest d = sha256.convert(randomKey);
    return EOSPrivateKey.fromBuffer(d.bytes);
  }

  /// Check if the private key is WIF format
  bool isWIF() => this.format == 'WIF';

  /// Get the public key string from this private key
  EOSPublicKey toEOSPublicKey() {
    BigInt privateKeyNum = decodeBigInt(this.d);
    ECPoint ecPoint = EOSKey.secp256k1.G * privateKeyNum;

    return EOSPublicKey.fromPoint(ecPoint);
  }

  /// Sign the string data using the private key
  EOSSignature sign(String data) {
    Digest d = sha256.convert(utf8.encode(data));
    return signHash(d.bytes);
  }

  /// Sign the SHA256 hashed data using the private key
  EOSSignature signHash(Uint8List sha256Data) {
    int nonce = 0;

    while (true) {
      ECSignature sig = _deterministicGenerateK(sha256Data, this.d, nonce);
      Uint8List der = EOSSignature.ecSigToDER(sig);

      int lenR = der.elementAt(3);
      int lenS = der.elementAt(5 + lenR);
      if (lenR == 32 && lenS == 32) {
        int i = EOSSignature.calcPubKeyRecoveryParam(
            decodeBigInt(sha256Data), sig, this.toEOSPublicKey());
        i += 4;
        i += 27;
        return EOSSignature(i, sig.r, sig.s);
      }
    }
  }

  String toString() {
    List<int> version = List<int>();
    version.add(EOSKey.VERSION);
    Uint8List keyWLeadingVersion =
    EOSKey.concat(Uint8List.fromList(version), this.d);

    return EOSKey.encodeKey(keyWLeadingVersion, EOSKey.SHA256X2);
  }

  ECSignature _deterministicGenerateK(Uint8List hash, Uint8List x, int nonce) {
    List<int> data = List();
    data.addAll(hash);
    data.add(nonce);

    Uint8List newHash = Uint8List.fromList(data);
    final signer = ECDSASigner(null, HMac(SHA256Digest(), 64));
    var pkp = new PrivateKeyParameter(
        ECPrivateKey(decodeBigInt(x), EOSKey.secp256k1));
    signer.init(true, pkp);
    return signer.generateSignature(newHash);
  }
}
